// jest-trx-results-processor/dist/utils.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.getEnvInfo = exports.getSuitePerTestDuration = exports.getTestClassName = exports.getFullTestName = exports.formatDuration = void 0;
var os = __importStar(require("os"));
// Adapted from https://github.com/hatchteam/karma-trx-reporter
exports.formatDuration = function (duration) {
    var durationInner = duration || 0;
    var ms = durationInner % 1000;
    durationInner -= ms;
    var s = (durationInner / 1000) % 60;
    durationInner -= s * 1000;
    var m = (durationInner / 60000) % 60;
    durationInner -= m * 60000;
    var h = (durationInner / 3600000) % 24;
    durationInner -= h * 3600000;
    var d = durationInner / 86400000;
    return ((d > 0 ? d + "." : "") +
        (h < 10 ? "0" + h : h) +
        ":" +
        (m < 10 ? "0" + m : m) +
        ":" +
        (s < 10 ? "0" + s : s) +
        "." +
        (ms < 10 ? "00" + ms : ms < 100 ? "0" + ms : ms));
};
// Auxillary test data functions
exports.getFullTestName = function (testResult) {
    return testResult.ancestorTitles && testResult.ancestorTitles.length
        ? testResult.ancestorTitles.join(" / ") + " / " + testResult.title
        : testResult.title;
};
exports.getTestClassName = function (testResult) {
    return testResult.ancestorTitles && testResult.ancestorTitles.length
        ? testResult.ancestorTitles[0]
        : "No suite";
};
exports.getSuitePerTestDuration = function (testSuiteResult) {
    // take the total duration of suite and divide it by the number of tests
    // (Jest does not provide per test performance info)
    return Math.floor((testSuiteResult.perfStats.end - testSuiteResult.perfStats.start) /
        (testSuiteResult.numPassingTests +
            testSuiteResult.numFailingTests +
            testSuiteResult.numPendingTests +
            testSuiteResult.numTodoTests));
};
exports.getEnvInfo = function (defaultUserName) {
    if (defaultUserName === void 0) { defaultUserName = "anonymous"; }
    return ({
        computerName: os.hostname(),
        userName: process.env.SUDO_USER ||
            process.env.LOGNAME ||
            process.env.USER ||
            process.env.LNAME ||
            process.env.USERNAME ||
            defaultUserName,
    });
};
